### PR TITLE
Fix validation code not sending on initial page load.

### PR DESCRIPTION
**Pull request by wildan-m**
### Details
Fix initial code validation in contact method and 2FA screen for PR.

### Fixed Issues

$ https://github.com/Expensify/App/issues/43359
PROPOSAL: https://github.com/Expensify/App/issues/43359#issuecomment-2304159073

### Tests

1. Login to as a new user
2. Go to profile settings --> Contact Methods
3. Wait for the magic code to arrive
4. Press Back <
5. Re-Select Contact Method
6. Verify that magic code only sent once
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Login to as a new user
2. Make network offline
3. Go to profile settings --> Contact Methods
4. Make network online
5. Wait for the magic code to arrive
6. Press Back <
7. Re-Select Contact Method
8. Verify that magic code only sent once
- [x] Verify that no errors appear in the JS console

### QA Steps

Same as Test
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/50a09623-fcc4-4d33-be49-4042bf5a9f31

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/c319807a-ec51-4c9a-a7dd-cf54041a481e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/f92d847d-58af-44d6-b2c3-1d765dcf0b30

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/555b2a70-fe15-425c-9957-89da5daeae02

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/0c10343e-b4a2-4c33-a9ae-478785ac4cb5

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/872120cf-5ab0-46fe-a0fb-a8fc32e932a4

</details>

### CODE REVIEW

**Comment by mountiny** [comment]
@wildan-m there are conflicts

@allroundexperts what is your eta for the review?

**Comment by wildan-m** [comment]
The conflict is resolved

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by allroundexperts** [comment]
@wildan-m If I repeat the steps, I can get the validation messages again:

https://github.com/user-attachments/assets/19687812-76d0-4107-bb61-391ae55f8bf0

**Comment by wildan-m** [comment]
@allroundexperts I added test steps 4-6 that seem reasonable, but this code intentionally resets the validateCodeSent state on each mount

https://github.com/Expensify/App/blob/8870d602c96294bf0d07657cf4c6cf90c6c7e241/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx#L139

Depending on the expected outcome, we can either eliminate the extra test steps or delete [this code](https://github.com/Expensify/App/blob/8870d602c96294bf0d07657cf4c6cf90c6c7e241/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx#L139).

**Comment by wildan-m** [comment]
@mountiny What is the expected behavior when [revisiting the validation page](https://github.com/Expensify/App/pull/48047#issuecomment-2316508626)?

**Comment by wildan-m** [comment]
@allroundexperts @mountiny bump for confirmation.

https://github.com/Expensify/App/pull/48047#issuecomment-2316508626

Should we re-send on each page load?

**Comment by allroundexperts** [comment]
I think @mountiny would be able to answer that.

**Comment by mountiny** [comment]
Hmmm, I am not sure; I don't think it is best to call it every time the page is navigated. Could we check if the `login.validateCodeSent` is true and only if its not, then send it. 

Then require the user to click the request button

**Comment by wildan-m** [comment]
@mountiny Thanks for your confirmation. in that case we can remove [this useEffect](https://github.com/Expensify/App/blob/8870d602c96294bf0d07657cf4c6cf90c6c7e241/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.tsx#L139).

@allroundexperts the PR has been updated

**Comment by mountiny** [comment]
@wildan-m  this flow was not slightly changed so we require the magic code also for the primary's login

Can you retest this flow with latest main merged?

**Comment by wildan-m** [comment]
@mountiny I found flicker in re-sent message and noticeable in iOS, but it has been fixed. Currently conducting more tests and will update the video soon

**Comment by wildan-m** [comment]
@mountiny @allroundexperts The video with latest main has been updated. https://github.com/Expensify/App/pull/48047#issue-2488267865

File: src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (allroundexperts) [inline src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx:1]: Why is this change needed @wildan-m? This doesn't seem to be related to your changes.

Comment (wildan-m) [reply src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx:1]: @allroundexperts did you mean adding `validateCodeSentIsPressedRef` ? without that, the message `Link has been re-sent` would be shown on the initial load.

Comment (wildan-m) [reply src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx:1]: @allroundexperts 

Before: The `requestContactMethodValidateCode` function is not triggered upon the initial page load, resulting in the `validateCodeSent` variable not being updated. It is only updated when the `magicCodeNotReceived` button is clicked, allowing `validateCodeSent` to solely indicate whether a magic code has been sent. Additionally, `validateCodeSent` is reset to false each time the component is mounted.

After: The function `requestContactMethodValidateCode` is invoked on each page call if `validateCodeSent` is false. When this function is called, `validateCodeSent` is changed to true. On the initial page load of `BaseValidateCodeForm.tsx`, the message `Link has been re-sent` will be displayed. We cannot rely solely on `validateCodeSent` because its value is modified before the page call.

Comment (wildan-m) [reply src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx:1]: @allroundexperts bump https://github.com/Expensify/App/pull/48047#discussion_r1750999831

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Review by mountiny — APPROVED** [review]
@wildan-m @allroundexperts sorry for this taking a bit longer, the changes look good to me, we started to use the validation page for adding secondary login. Can you validate that adding secondary login works on this PR as well please?

**Comment by wildan-m** [comment]
@mountiny In my video, the secondary login test at the end seemed to work as expected.

**Comment by mountiny** [comment]
@wildan-m nice, I will also check the the reveal virtual card details flow works

**Comment by mountiny** [comment]
@wildan-m can you check the wallet page? unfortunately its hard for you to test but I am not getting the validation code on this branch to reveal the card details. That is the Reveal details on the card page
IMG_336A39EAAD24-1

**Comment by wildan-m** [comment]
@mountiny there are three  `BaseValidateCodeForm`

1. [signing](https://github.com/Expensify/App/blob/main/src/pages/signin/ValidateCodeForm/BaseValidateCodeForm.tsx)
2. [Contact](https://github.com/Expensify/App/blob/main/src/pages/settings/Profile/Contacts/ValidateCodeForm/BaseValidateCodeForm.tsx)
3. [ValidateCodeActionModal](https://github.com/Expensify/App/blob/main/src/components/ValidateCodeActionModal/ValidateCodeForm/BaseValidateCodeForm.tsx) (seems this is recently added)

We are currently focused on the second `BaseValidateCodeForm`, the first one seems not require change. The third one is not directly called from route but mounted in modal, in that case current solution will not working, we need to implement `useEffect` solution.

Should we include the recently added third item in this PR, even though it's a bit difficult to test? If so, how can we simplify the testing process?

**Review by grgia — APPROVED** [review]

**Comment by grgia** [comment]
> Should we include the recently added third item in this PR, even though it's a bit difficult to test? If so, how can we simplify the testing process?

Just saw this. I think we should- both @mountiny and I can test in the adhoc builds. 
Thoughts @allroundexperts @wildan-m ?

**Comment by mountiny** [comment]
@grgia I think we should do this in a follow up, this is already fixing existing bugs. I will ask in slack

**Comment by mountiny** [comment]
Started a thread [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1726483475882029)

File: src/libs/actions/User.ts [hunk 1]
```diff
@@ -412,6 +412,7 @@ function addNewContactMethod(contactMethod: string, validateCode = '') {
                 [contactMethod]: {
                     partnerUserID: contactMethod,
                     validatedDate: '',
+                    validateCodeSent: true,
```
[the remark this answers is not in the document]
Comment (wildan-m) [reply src/libs/actions/User.ts:415]: @getusha at the time PR created, when we call ADD_NEW_CONTACT_METHOD, BE will sent the validation message. Did you found any issue with the code? or BE behave differently at the moment?

[the remark this answers is not in the document]
Comment (wildan-m) [reply src/libs/actions/User.ts:415]: @getusha if the BE not sending validation now, I think you can safely remove that

[the remark this answers is not in the document]
Comment (wildan-m) [reply src/libs/actions/User.ts:415]: I think BE still behave the same

https://github.com/user-attachments/assets/dab8a876-990e-4928-9710-bea85d3dca16

[the remark this answers is not in the document]
Comment (wildan-m) [reply src/libs/actions/User.ts:415]: @getusha If you have a plan to delete it, make sure the resent message works properly. iOS is a bit troublesome with flickering, but I don't think it's directly related.
